clippy_lints/src/bool_assert_comparison.rs

```diff
@@ -0,0 +1,75 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::{ast_utils, is_direct_expn_of};
+use rustc_ast::ast::{Expr, ExprKind, Lit, LitKind};
+use rustc_errors::Applicability;
+use rustc_lint::{EarlyContext, EarlyLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::DUMMY_SP;
+
+declare_clippy_lint! {
+    /// **What it does:** This lint warns about boolean comparisons in assert-like macros.
+    ///
+    /// **Why is this bad?** It is shorter to use the equivalent.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// // Bad
+    /// assert_eq!("a".is_empty(), false);
+    /// assert_ne!("a".is_empty(), true);
+    ///
+    /// // Good
+    /// assert!(!"a".is_empty());
+    /// ```
+    pub BOOL_ASSERT_COMPARISON,
+    pedantic,
+    "Using a boolean as comparison value when there is no need"
+}
+
+declare_lint_pass!(BoolAssertComparison => [BOOL_ASSERT_COMPARISON]);
+
+impl EarlyLintPass for BoolAssertComparison {
+    fn check_expr(&mut self, cx: &EarlyContext<'_>, e: &Expr) {
+        let macros = ["assert_eq", "debug_assert_eq"];
+        let inverted_macros = ["assert_ne", "debug_assert_ne"];
+
+        for mac in macros.iter().chain(inverted_macros.iter()) {
+            if is_direct_expn_of(e.span, mac).is_some() {
+                if let Some(macro_args) = ast_utils::extract_assert_macro_args(e) {
+                    let mut nb_bool_args = 0;
+                    let mut bool_span = DUMMY_SP;
+                    for arg in macro_args {
+                        if let ExprKind::Lit(Lit {
+                            kind: LitKind::Bool(_), ..
+                        }) = arg.kind
+                        {
+                            nb_bool_args += 1;
+                            bool_span = arg.span;
+                        }
+                    }
+                    #[allow(clippy::if_same_then_else)] // It allows better code reability here.
+                    if nb_bool_args > 1 {
+                        // Don't know what's going on here, but better leave things as is...
+                        return;
+                    } else if nb_bool_args == 0 {
+                        // No problem detected!
+                        return;
+                    }
+
+                    span_lint_and_sugg(
+                        cx,
+                        BOOL_ASSERT_COMPARISON,
+                        bool_span,
+                        "assert macro with a boolean comparison",
```

Comment:
The "literal" part is important.

```suggestion
                        format!("used {}! with a literal bool", mac),
```

Reply:
:+1: